assembl/graphql/schema.py

```diff
@@ -387,14 +608,265 @@ def mutate(root, args, context, info):
+        if not allowed or (allowed == IF_OWNED and user_id == Everyone):
+            raise HTTPUnauthorized()
+
+        thematic.tombstone_date = datetime.utcnow()
+        thematic.db.flush()
+        return DeleteThematic(success=True)
+
+
+class CreatePost(graphene.Mutation):
+    class Input:
+        subject = graphene.String()
+        body = graphene.String(required=True)
+        idea_id = graphene.ID(required=True)
+        # TODO: for a real post (not a proposal), we need to handle parent_id
+        # see related code in views/api/post.py
+
+    post = graphene.Field(lambda: PostUnion)
+
+    @staticmethod
+    def mutate(root, args, context, info):
+        discussion_id = context.matchdict['discussion_id']
+
+        user_id = context.authenticated_userid or Everyone
+
+        idea_id = args.get('idea_id')
+        idea_id = int(Node.from_global_id(idea_id)[1])
+        in_reply_to_idea = models.Idea.get(idea_id)
+        if isinstance(in_reply_to_idea, models.Question):
+            cls = models.PropositionPost
+        else:
+            cls = models.Post
+
+        permissions = get_permissions(user_id, discussion_id)
+        allowed = cls.user_can_cls(user_id, CrudPermissions.CREATE, permissions)
+        if not allowed or (allowed == IF_OWNED and user_id == Everyone):
+            raise HTTPUnauthorized()
+
+        with cls.default_db.no_autoflush:
+            subject = args.get('subject')
+            body = args.get('body')
+            if subject is None:
+                # TODO for real post, need the same logic than in views/api/post.py
+                subject_entries = [
+                    {'value': u'Proposition', u'locale_code': u'und'}
+                ]
+            else:
+                subject_entries = [
+                    {'value': subject, u'locale_code': u'und'}
+                ]
+
+            body_entries = [
+                {'value': body, u'locale_code': u'und'}
+            ]
+
+            subject_langstring = langstring_from_input_entries(subject_entries)
+            body_langstring = langstring_from_input_entries(body_entries)
+            discussion = models.Discussion.get(discussion_id)
+            new_post = cls(
+                discussion=discussion,
+                subject=subject_langstring,
+                body=body_langstring,
+                creator_id=user_id
+            )
+            db = new_post.db
+            db.add(new_post)
+            idea_post_link = models.IdeaRelatedPostLink(
+                creator_id=user_id,
+                content=new_post,
+                idea=in_reply_to_idea
+            )
+            db.add(idea_post_link)
+            db.flush()
+
+        return CreatePost(post=new_post)
+
+
+class AddSentiment(graphene.Mutation):
+    class Input:
+        post_id = graphene.ID(required=True)
+        type = graphene.Argument(
+            type=SentimentTypes,
+            required=True
+        )
+
+    post = graphene.Field(lambda: PostUnion)
```

Comment:
Yes, this is actually to avoid a cyclic reference. For example if you wrote graphene.Field(PostUnion) and PostUnion wasn't defined above, you get and error. Sometime it's not possible to put the classes you use above because of a cyclic use.